src/bitops.c

```diff
@@ -697,81 +891,173 @@ void bitopCommand(client *c) {
          * can take a fast path that performs much better than the
          * vanilla algorithm. On ARM we skip the fast path since it will
          * result in GCC compiling the code using multiple-words load/store
          * operations that are not supported even in ARM >= v6. */
-        j = 0;
-        #ifndef USE_ALIGNED_ACCESS
-        if (minlen >= sizeof(unsigned long)*4 && numkeys <= 16) {
-            unsigned long *lp[16];
+        if (minlen >= sizeof(unsigned long)*4) {
+            unsigned long **lp = (unsigned long**)src;
             unsigned long *lres = (unsigned long*) res;
+            unsigned long *first_key = (unsigned long*)src[0];
+            unsigned long lcommon_bits[4];
 
-            memcpy(lp,src,sizeof(unsigned long*)*numkeys);
-            memcpy(res,src[0],minlen);
+            const size_t step = sizeof(unsigned long)*4;
+            size_t processed = 0;
+            size_t k = 0;
+
+            if (op != BITOP_DIFF && op != BITOP_DIFF1 && op != BITOP_ANDOR)
+                memcpy(lres,src[0],minlen);
 
             /* Different branches per different operations for speed (sorry). */
             if (op == BITOP_AND) {
-                while(minlen >= sizeof(unsigned long)*4) {
+                while(minlen >= step) {
                     for (i = 1; i < numkeys; i++) {
-                        lres[0] &= lp[i][0];
-                        lres[1] &= lp[i][1];
-                        lres[2] &= lp[i][2];
-                        lres[3] &= lp[i][3];
-                        lp[i]+=4;
+                        lres[0] &= lp[i][k+0];
+                        lres[1] &= lp[i][k+1];
+                        lres[2] &= lp[i][k+2];
+                        lres[3] &= lp[i][k+3];
                     }
-                    lres+=4;
-                    j += sizeof(unsigned long)*4;
-                    minlen -= sizeof(unsigned long)*4;
+                    k += 4;
+                    lres += 4;
+                    j += step;
+                    minlen -= step;
                 }
             } else if (op == BITOP_OR) {
-                while(minlen >= sizeof(unsigned long)*4) {
+                while(minlen >= step) {
                     for (i = 1; i < numkeys; i++) {
-                        lres[0] |= lp[i][0];
-                        lres[1] |= lp[i][1];
-                        lres[2] |= lp[i][2];
-                        lres[3] |= lp[i][3];
-                        lp[i]+=4;
+                        lres[0] |= lp[i][k+0];
+                        lres[1] |= lp[i][k+1];
+                        lres[2] |= lp[i][k+2];
+                        lres[3] |= lp[i][k+3];
                     }
-                    lres+=4;
-                    j += sizeof(unsigned long)*4;
-                    minlen -= sizeof(unsigned long)*4;
+                    k += 4;
+                    lres += 4;
+                    j += step;
+                    minlen -= step;
```

Comment:
Are these codes just refactoring?

Reply:
@sundb I can maybe return `+= sizeof(unsigned int) * 4` if that'll be clearer and instead of `j += ...` we could say `j = k * sizeof(unsigned int)` at the end of the if-else chain.